Add OCaml support

huwaireb opened the pull request:
👋 

This pull request implements support for the [OCaml Language](https://ocaml.org/). 

### Additions
- [x] [tree-sitter-ocaml](https://github.com/tree-sitter/tree-sitter-ocaml) grammar
- [x] Highlight, Indents, Outline queries
- [x] A new file icon for .ml(i) files. Based on [ocaml/ocaml-logo](https://github.com/ocaml/ocaml-logo/blob/master/Colour/SVG/colour-transparent-icon.svg)
- [x] LSP Integration with [ocaml-language-server](https://github.com/ocaml/ocaml-lsp)
- [x] Completion Labels
- [x] Symbol Labels

### Bug Fixes
- [x] Improper parsing of LSP headers. 

### Missing [will file a separate issue]
- Documentation on completionItem, requires: `completionItem/resolve` with support for `documentation` as a provider. See #7288 

### Screenshots

<details><summary>Zed</summary>
<img width="1800" alt="Screenshot 2024-02-01 at 03 33 20" src="https://github.com/zed-industries/zed/assets/69181766/e17c184e-203e-40c3-a08f-4de46226b79c">
</details>

Release Notes:
- Added OCaml Support ([#5316](https://github.com/zed-industries/zed/issues/5316)).

> [!NOTE]
> Partially completes #5316 
> To complete #5316:
> 1. addition of a reason tree-sitter grammar.
> 2. opam/esy integration, however it may be better as it's own plugin.

huwaireb committed:
feat: support tree-sitter ocaml

huwaireb committed:
feat: init ocamllsp

huwaireb committed:
feat: add ocaml icon

huwaireb commented:
outline queries need improvements because it currently outlines the following

```ocaml
let A =  ... @item
    let module B = ... in @item
	let C = ... in  @item
	...
```

a solution could easily be implemented by scoping the value_definition queries under (compilation_unit), but that would also mean outline wouldn't include scoped definitions e.g

```ocaml
module A = struct
	let C : ... @item
end
```

should i duplicate the queries or is there a better way?

huwaireb commented:
@cla-bot check

huwaireb commented:
outline queries need improvements because it currently outlines the following

```ocaml
let A =  ... @item
    let module B = ... in @item
	let C = ... in  @item
	...
```

a solution could easily be implemented by scoping the value_definition queries under (compilation_unit), but that would also mean outline wouldn't include scoped definitions e.g

```ocaml
module A = struct
	let C : ... @item
end
```

should i duplicate the queries or is there a better way?

huwaireb commented:
The indents and outlines are working great now,

there is just one minor issue with the indents, they don't indent till i write something:

https://github.com/zed-industries/zed/assets/69181766/35dd0dc4-0699-491a-aef9-24dfa76659b0

but oddly they do work well if scoped under a module

https://github.com/zed-industries/zed/assets/69181766/8e8dfd6f-0c84-4c1c-ad51-7b479e5b3da6

not the slightest idea how to fix this.

huwaireb committed:
fix(crate/lsp): handle optional header field Content-Type

huwaireb commented:
I've pushed a commit to handle Content-Type, the change felt too small for a new PR. The LSP should work fine now.

SomeoneToIgnore commented:
Hello, could you also add a docs entry please, into the https://github.com/zed-industries/zed/tree/main/docs/src/languages list.

huwaireb committed:
docs: add ocaml to languages

huwaireb commented:
For some reason, logic to handle resolving documentation for Completionitem was there, just not included in the resolve_providers of the completion items. This caused no documentation to show for CompletionItem's, not sure why it was omitted, but it seems to work fine?

I think it should be supported nonetheless based on the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion):

> If computing full completion items is expensive, servers can additionally provide a handler for the completion item resolve > request (‘completionItem/resolve’.  A typical use case is for example: the textDocument/completion request doesn’t fill in > > the documentation property for returned completion items since it is expensive to compute

In eithercase, I pushed a change if it's fine, otherwise will revert.

huwaireb commented on crates/lsp/src/lsp.rs:
```diff
@@ -540,7 +540,7 @@ impl LanguageServer {
                             snippet_support: Some(true),
                             resolve_support: Some(CompletionItemCapabilityResolveSupport {
-                                properties: vec!["additionalTextEdits".to_string()],
+                                properties: vec!["documentation".to_string(), "additionalTextEdits".to_string()],
```
For some reason, logic to handle resolving documentation for Completionitem was there, just not included in the resolve_providers of the completion items. This caused no documentation to show for CompletionItem's for LSPs that require resolving completionItems for documentation. 

Not sure why it was omitted as it seems to work fine? Should be supported nonetheless based on the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion):

> If computing full completion items is expensive, servers can additionally provide a handler for the completion item resolve request (‘completionItem/resolve’.  A typical use case is for example: the textDocument/completion request doesn’t fill in the documentation property for returned completion items since it is expensive to compute

In eithercase, I pushed a change if it's fine, otherwise will revert.

huwaireb replied:
Work fine was something I should've tested, it doesn't. I've removed this commit. Unfortunately it would mean lack of documentation on CompletionItem's, but better than to have a buggy experience..

huwaireb commented:
![image](https://github.com/zed-industries/zed/assets/69181766/c7517175-cd0c-4c9c-a0fa-fc91cca62cd8)

I noticed documentation highlighting is broken, not Zed's fault, but ocamllsp for sending invalid ocaml code. Probably due to testing on textmate grammars. A fix is so simple it's just prepending `module NAME :`, but there doesn't seem to be any process_documentation function available, so probably needs to be done in ocamllsp.

this seems to also be a problem in other tree-sitter based editors e.g helix.

huwaireb committed:
feat: support completion labels

huwaireb committed:
feat: support symbol labels for ocamllsp

huwaireb commented:
rebased, should be ready now 🪷

huwaireb commented:
I've thought this was a zed completion menu bug, 

<img width="734" alt="Screenshot 2024-02-01 at 09 48 56" src="https://github.com/zed-industries/zed/assets/69181766/f877ec38-3e13-4319-9d32-2e44ab290b08">

but turns out it was because ocamllsp was returning \n for some completion item details for some reason. Not sure if i should handle it on Zed's side or seek a solution in ocamllsp.

huwaireb committed:
fix(ocamllsp): replace any line breaks in completion detail with whitespace

huwaireb commented:
I'm curious if it's fine for me to document use instructions across different package managers (nix/esy/opam) in the`docs/src/languages`? 

The lack of direct support for the officially supported package manager (opam) would require a similar workflow to using Nix with any other language. That is, opening Zed from a terminal where your active project sandbox (switch) is activated with ocamllsp installed.

I would've loved to implement it into Zed to make it more seamless, but I guess that would set precedent for other "obscure" languages that it's fine, as opposed to going through the soon to be plugin route.

SomeoneToIgnore commented on docs/src/languages/ocaml.md:
```diff
@@ -0,0 +1,4 @@
+# OCaml
```
I see that there's no mac artifact to download from https://github.com/ocaml/ocaml-lsp/releases which means users have to set up ocaml LSP add add to the path themselves, let's mention it in the docs here.

huwaireb replied:
I've added some instructions, please give them a shot!

SomeoneToIgnore commented on crates/zed/src/languages/ocaml/config.toml:
```diff
@@ -0,0 +1,17 @@
+  { start = "begin", end = "end", close = true, newline = true },
+  { start = "struct", end = "end", close = true, newline = true },
+  { start = "sig", end = "end", close = true, newline = true },
+  { start = "object", end = "end", close = true, newline = true },
```
FYI: I was never able to get the `object` closed with `end` in the code, but was able to do that with `sig`

I also think all word-like "brackets" would benefit from some space in the start of the closing part, wdyt?
![image](https://github.com/zed-industries/zed/assets/2690773/0efd2843-95cb-4018-a229-f9aa6f9570d3)

huwaireb replied:
1. So i'm testing this now, and it seems like having start as "object" might be a reserved word as anything else e.g "objec" works fine. Not sure how to fix this.. A temporary hack could be to do it as "object "

2. Yes, that sounds great. I'll prepend whitespace before each "end".

huwaireb replied:
I've partially fixed 1 with "object " for now. It'll require you to type object with a space but it's better than nothing i guess.

done 2.

SomeoneToIgnore replied:
Indeed, thanks for taking care of that.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -332,11 +332,33 @@ impl LanguageServer {
+
+            let message_len: usize = segments
+                .next()
+                .ok_or_else(|| anyhow!("unable to find the first line of the LSP message header"))? // Impossible, but just in case.
```
Let's remove `ok_or_else|| anyhow!(` with `.context("` calls here and below.
`return Err(anyhow!` could be `anyhow::bail!` too, for readability.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -332,11 +332,33 @@ impl LanguageServer {
+                .parse()
+                .context("failed to parse Content-Length of LSP message header")?;
+
+            if let Some(second_segment) = segments.next() {
```
Curious to know the reason for changes here: was it the first ever server to actually send content-type to us?

huwaireb replied:
Yep, unfortunately was! see #7220 for another LSP that does it as well.

SomeoneToIgnore replied:
Oh, thank you for biting the bullet for us and fixing it.

huwaireb replied:
I'm not handling the actual value of the Content-Type however, is it fine? 

Seemed useless as they wouldn't have gotten past certain points if they were anything but application/vscode-jsonrpc and utf(-)8.

SomeoneToIgnore replied:
Given that we did not handle anything before that and were fine — sure, you can leave it as is.
Some follow-up PRs on the topic are welcome, but seems not necessary yet too.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -332,11 +332,33 @@ impl LanguageServer {
-                .trim_end()
-                .parse()?;
+                .parse()
+                .context("failed to parse Content-Length of LSP message header")?;
```
Let's use `with_context` and actually print the erroneous header in the message.

SomeoneToIgnore replied:
This one still on, though.
Sorry for nitpicking, but here you change all LSP-related code, so let's do one more round and actually use `.with_context(|| format!(...` and mention what header exactly did we fail to parse.

SomeoneToIgnore reviewed (commented):
Nice work here, thanks!

I could not test it without the server installed, but looks like it works for you and has such great *.scm files already, so let's fix the remaining nits and merge this.

huwaireb committed:
address pull request review

Cleans up LSP header parsing
Prepends whitespace to end and done
Cleans up indentation rules for ocaml-interface
Adds setup instructions for OCaml

SomeoneToIgnore commented:
Go test started to fail after a different commit in `main` and I'll check that, but this PR has nothing to do with it, optimistically merging.

Thank you for the fixes!
